src/hotspot/share/gc/g1/g1CollectedHeap.cpp

```diff
@@ -2093,6 +2093,38 @@ bool G1CollectedHeap::try_collect_concurrently(GCCause::Cause cause,
   }
 }
 
+bool G1CollectedHeap::try_collect_fullgc(GCCause::Cause cause,
+                                         const G1GCCounters& counters_before) {
+  assert_heap_not_locked();
+
+  while(true) {
+    VM_G1CollectFull op(counters_before.total_collections(),
+                        counters_before.total_full_collections(),
+                        cause);
+    VMThread::execute(&op);
+
+    // Request is trivially finished.
+    if (op.gc_succeeded() || cause == GCCause::_g1_periodic_collection) {
+      return op.gc_succeeded();
+    }
+
+    uint full_gc_count = 0;
+    {
+      MutexLocker ml(Heap_lock);
+      full_gc_count = total_full_collections();
+    }
+
+    if (counters_before.total_full_collections() != full_gc_count) {
+      return true;
+    }
```

Comment:
The other two GCs does this check inside the locker-scope. I don't think there's any practical diff -- why the inconsistency?

Reply:
Only difference is "time" at which changes were made (Initially on G1), then also I couldn't decide which version is cleaner.

Reply:
I think the style in Serial is cleaner.